src/commons/listener.rs

```diff
@@ -103,6 +133,56 @@ pub struct ListenerStatus {
 pub struct ListenerIngress {
     /// The hostname or IP address to the [`Listener`].
     pub address: String,
+    /// The type of address (`Hostname` or `IP`).
+    pub address_type: AddressType,
```

Comment:
We could maybe combine those into a complex enum `ListenerAddress`:

```rust
pub enum ListenerAddress {
    Hostname(String),
    Ip(std::net::IpAddress),
}
```

Reply:
Yes, I was tempted to do this (and we already do this internally in secret-op). The downside is that it means that we need to either always `#[serde(flatten)]` it in, or introduce another level of nesting (breaking compatibility).

Reply:
I'm also not sure about a good naming scheme in that case.. then again, now would be the time, if ever, to do that compat break.

Reply:
Yeah I feel like the address should _itself_ contain the type/kind of address it is. Keeping track of this in two different fields seems a little weird.

Reply:
Regarding the naming: I think `ListenerAdress` seems nice, but `ListenerIngressAddress` might technically be more correct.

Reply:
§orry, in terms of naming I mean… `address.type` makes sense, but `address.address` feels very redundant. I'd like to keep it in the same field rather than split it into `address.ip` and `address.hostname`, since many consumers should be fine with either…

Reply:
Sorry, I don't quite follow. The structure would look like this:

```rust
pub struct ListenerIngress {
    pub address: ListenerIngressAddress,
    pub ports: BTreeMap<String, i32>, // I would use u32, as there are no negative ports
}

pub enum ListenerIngressAddress {
    Ip(std::net::IpAddress),
    Hostname(String),
}
```

The address could then be retrieved by `ingress.address`. We then could use match to handle both variants.

Reply:
I think that @Techassi's proposal is to derive the address type from the content of the address field instead of explicitly defining it:

```yaml
# no address type field
address: 10.1.2.3                 # looks like an IPv4 address
address: 2001:db8::8a2e:370:7334  # looks like an IPv6 address
address: example.com              # looks like a hostname
```

In my opinion, we should imitate the Kubernetes API in our resource definitions as close as possible so that the user is not confronted with another Kubernetes lingo. For instance, an EndpointSlice looks as follows:

```yaml
apiVersion: discovery.k8s.io/v1
kind: EndpointSlice
addressType: IPv4
endpoints:
  - addresses:
      - 10.1.2.3
```

The `addressType` (`IPv4`, `IPv6`, `FQDN`) specifies the content of the `addresses` field.

The current implementation of the `ListenerIngress` with `address` and `addressType` is close to it. So I would leave it as is.

Reply:
I tried to reply to this the other day, but it seems to have been eaten by the comment monster. Basically, my concern is that I specifically _don't_ want people to have to pattern match if they just want the address, since the common case "should" be to just use it in a URL or whatever.